File: 2872-maximum-number-of-k-divisible-components/2872-maximum-number-of-k-divisible-components.cpp
#pragma GCC optimize("O3", "unroll-loops");
static const int __ = []() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    return 0;
}();
using ll = long long;
class Solution {
   public:
    int ans = 0;
    int k;
    vector<vector<int>> adj;
    ll dfs(int node, int parent, vector<int>& values) {
        ll currentSum = values[node];
        for (int neighbor : adj[node]) {
            if (neighbor != parent) {
                currentSum += dfs(neighbor, node, values);
            }
        }

        if (currentSum % k == 0) {
            ans++;
            return 0;
        }
        return currentSum;
    }
    int maxKDivisibleComponents(int n, vector<vector<int>>& edges,
                                vector<int>& values, int k) {
        adj.resize(n);
        this->k = k;

        for (const auto& edge : edges) {
            int u = edge[0];
            int v = edge[1];
            adj[u].emplace_back(v);
            adj[v].emplace_back(u);
        }

        dfs(0, -1, values);

        return ans;
    }
};

File: 2872-maximum-number-of-k-divisible-components/2872-maximum-number-of-k-divisible-components.rs
impl Solution {
    pub fn max_k_divisible_components(
        n: i32,
        edges: Vec<Vec<i32>>,
        values: Vec<i32>,
        k: i32,
    ) -> i32 {
        let n = n as usize;
        let mut adj = vec![vec![]; n];
        let mut ans = 0;
        for edge in edges {
            let u = edge[0] as usize;
            let v = edge[1] as usize;
            adj[u].push(v);
            adj[v].push(u);
        }

        fn dfs(
            curr: usize,
            parent: usize,
            values: &Vec<i32>,
            adj: &Vec<Vec<usize>>,
            k: i64,
            ans: &mut i32,
        ) -> i64 {
            let mut current_sum = values[curr] as i64;

            for &neighbor in &adj[curr] {
                if neighbor != parent {
                    current_sum += dfs(neighbor, curr, values, adj, k, ans);
                }
            }

            if current_sum % k == 0 {
                *ans += 1;
                return 0;
            }

            current_sum
        }

        dfs(0, usize::MAX, &values, &adj, k as i64, &mut ans);

        ans
    }
}